### DISCUSSION TITLE
Locally using the build files on real mobile phones

### DISCUSSION THREAD

**Discussion by emircetinmemis** [codemagic-ci-cd/codemagic-docs#2311] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

1. Prepare a project
2. Build it via codemagic
3. Download artifact files (android & ios)
4. Now, I don't know what to do.

### Expected results

Somehow, I want to dowload these artifacts, then I mean directly send them to my phone for example and run it locally.

### Actual results

I only have files on my windows machine's desktop currently. I want to test the ios build on my real iphone device.

### Build id (optional)

65a687c452ee5e479368103f

**Comment by dtrdic** [comment]
Hello @emircetinmemis, all iOS applications have to be digitally signed before they can be installed on real devices, and signing iOS applications requires Apple Developer Program membership. 
So, if you want to test your app on a real device check [this guide](https://docs.codemagic.io/flutter-code-signing/ios-code-signing/) on how to set up iOS code signing in the first place.
Check also this guide; https://labs.codemagic.io/your-first-flutter-app-to-appstore/#3 which will help you set up all the requirements on the AppStore as well

**Reply by dtrdic** [reply]
as for the android version, you can still test it on your device, just choose the `Android app bundle and universal APK` in the Android build format dropdown
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/38918425/2e773e09-e6c1-4ad4-83c7-22a2de2f57c8)

and when the artifact is created, you'll notice a QR code icon next to your .apk artifact. This icon provides a convenient way to quickly download and install the app on your device via the provided download link
<img width="407" alt="image" src="https://github.com/codemagic-ci-cd/codemagic-docs/assets/38918425/584eab88-6264-49ef-b6a4-dd3a331d7cdf">

**Comment by emircetinmemis** [comment]
Hi, there is a problem. That's why I feel like I need to reopen this issue instead of creating a new one.

I have been trying codemagic for a couple of days. But I'm just having a really hard time to proccess.

More or less, I took some progress, and I just followed your document below;
https://labs.codemagic.io/your-first-flutter-app-to-appstore/#5

And now, I did everything as expected in the given documentation and see and error as follows;
```
Failure: POST https://api.appstoreconnect.apple.com/v1/betaAppReviewSubmissions returned 422: Beta contract is missing for the app. - Beta Contract is missing.
```

Also might be important, I see some errors on publishing part also such as;
```
stdout: (null)
2024-01-20 22:19:16.505 *** Error: /Applications/Xcode-15.1.app/Contents/Developer/usr/bin/bitcode-build-tool, terminated with status: 2
2024-01-20 22:19:16.505 *** Error: /Applications/Xcode-15.1.app/Contents/Developer/usr/bin/bitcode-build-tool (2): bitcode-build-tool internal error
```

And also, when I try to open the .ipa file with my phone or throught testflight it is just not working not even installing it.

I have also faced with other errors at the beginning, but to make sure that this is not my fault, I just created a dummy init project and test on it.

Note that this is a new app because I tried on a dummy fresh one.
`Build ID : 65ac463118a21842341111ac`

For any additional information, I'm eager to wait for your responses.

**Reply by dtrdic** [reply]
Resolved over Discord. In conclusion, the issue was addressed by contacting Apple Support, and the app started working suddenly. However, there was no feedback or information received from Apple Support regarding the resolution
